web_search/perplexity.py

```diff
@@ -0,0 +1,107 @@
+#
+# perplexity.py
+#
+# Web search tool implementation based on Perplexity. Cannot perform searches with images.
+#
+
+from typing import Any, Dict, List
+
+import aiohttp
+from pydantic import BaseModel
+
+from web_search import WebSearch, WebSearchResult
+from models import Role, Message, TokenUsage, accumulate_token_usage
+
+class PerplexityMessage(BaseModel):
+    role: str = None
+    content: str = None
+
+class MessageChoices(BaseModel):
+    index: int = None
+    finish_reason: str = None
+    message: PerplexityMessage = None
+    delta: dict = None
+
+class Usage(BaseModel):
+    prompt_tokens: int = None
+    completion_tokens: int = None
+    total_tokens: int = None
+
+class PerplexityResponse(BaseModel):
+    id: str = None
+    model: str = None
+    created: int = None
+    usage: Usage = None
+    object: str = None
+    choices: List[MessageChoices] = None
+    
+    def summarise(self) -> str:
+        if len(self.choices) > 0:
+            return self.choices[0].message.content
+        else:
+            return "No results"
+
+class PerplexityWebSearch(WebSearch):
+    def __init__(self, api_key: str):
+        super().__init__()
+        self._api_key = api_key
+        self._session = None
+
+    def __del__(self):
+        if self._session:
+            self._session.detach()
+
+    async def _lazy_init(self):
+        if self._session is None:
+            # This instantiation must happen inside of an async event loop
+            self._session = aiohttp.ClientSession()
+    
+    async def search_web(self, query: str, token_usage_by_model: Dict[str, TokenUsage], use_photo: bool = False, image_bytes: bytes | None = None, location: str | None = None) -> WebSearchResult:
+        await self._lazy_init()
+        model = "pplx-7b-online"
```

Comment:
can we pass model a parameter? , as perplexity allows different model

Reply:
Done. Also made Perplexity default.